### PR TITLE
Tree of Thought introducing a new ToTChain.

### PR DESCRIPTION
**Pull request by vadimgu**
# [WIP] Tree of Thought introducing a new ToTChain.

This PR adds a new chain called ToTChain that implements the ["Large Language Model Guided Tree-of-Though"](https://arxiv.org/pdf/2305.08291.pdf) paper.

There's a notebook example `docs/modules/chains/examples/tot.ipynb` that shows how to use it.

Implements #4975

## Who can review?

Community members can review the PR once tests pass. Tag maintainers/contributors who might be interested:

- @hwchase17
- @vowelparrot

### CODE REVIEW

File: langchain/chains/tot/base.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+                color="red",
+            )
+
+    def solve(
+        self,
+        problem_description: str,
+        run_manager: Optional[CallbackManagerForChainRun] = None,
+    ) -> Optional[str]:
+        """Algorithm 2 from the ToT paper."""
+        checker = partial(self.checker, problem_description)
```
Comment (hwchase17) [inline langchain/chains/tot/base.py:273]: nit: not sure these lines add much

File: langchain/chains/tot/base.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        llm = self.llm.predict
+        controller = self.tot_controller
+        prompter = self.tot_prompter
+        memory = self.tot_memory
+
+        prompt = prompter(problem_description)
+        for _ in range(self.k):
+            self.log_request(prompt, run_manager)
+            response = self.parse_llm_output(llm(prompt))
```
Comment (hwchase17) [inline langchain/chains/tot/base.py:282]: we should construct and use LLMChain here - will help with logging (may make log_request and log_response unneccessary

**Review by hwchase17 — COMMENTED** [review]
awesome! left some stylistic comments, the one about using LLMChain is the main one

is there a better example? right now in the notebook seems to only use one iteration?

File: langchain/chains/tot/base.py (hunk 1 above)
Comment (vadimgu) [reply langchain/chains/tot/base.py]: I was trying to write something that looked very close to the algorithm in the paper. But I agree that it doesn't make it more readable.

**Comment by vadimgu** [comment]
I test it with Python3.8 and removed a listing issue with an unused import.

**Comment by vadimgu** [comment]
Do you think that ToTChecker should be a chain?

**Comment by levalencia** [comment]
is this already in the latest available version?

**Pull request merged**